src/operations/search/search.js

```diff
@@ -207,108 +208,162 @@ module.exports.search = async (requestInfo, args, resourceName, collection_name)
+        const useTwoStepSearchOptimization = !args['_elements'] && !args['id']
+            && (isTrue(env.USE_TWO_STEP_SEARCH_OPTIMIZATION) || args['_useTwoStepOptimization']);
+        if (useTwoStepSearchOptimization) {
+            // first get just the ids
+            const projection = {};
+            projection['_id'] = 0;
+            projection['id'] = 1;
+            options['projection'] = projection;
+            originalQuery = [query];
+            originalOptions = [options];
+            let idResults = await collection.find(query, options).maxTimeMS(maxMongoTimeMS).toArray();
+            if (idResults.length > 0) {
+                // now get the documents for those ids.  We can clear all the other query parameters
+                query = {'id': {$in: idResults.map(r => r.id)}};
+                // query = getQueryWithSecurityTags(securityTags, query);
+                options = {}; // reset options since we'll be looking by id
+                originalQuery.push(query);
+                originalOptions.push(options);
+            } else { // no results
+                query = null; //no need to query
             }
         }
 
-        // if _total is specified then ask mongo for the total else set total to 0
-        let total_count = 0;
-        if (args['_total'] && (['accurate', 'estimate'].includes(args['_total']))) {
-            // https://www.hl7.org/fhir/search.html#total
-            // if _total is passed then calculate the total count for matching records also
-            // don't use the options since they set a limit and skip
-            if (args['_total'] === 'estimate') {
-                total_count = await collection.estimatedDocumentCount(query, {maxTimeMS: maxMongoTimeMS});
-            } else {
-                total_count = await collection.countDocuments(query, {maxTimeMS: maxMongoTimeMS});
-            }
-        }
-        // Resource is a resource cursor, pull documents out before resolving
         /**
          * resources to return
          * @type {Resource[]}
          */
         let resources = [];
-        while (await cursor.hasNext()) {
+        /**
+         * @type {number}
+         */
+        let total_count = 0;
+        /**
+         * which index hint to use (if any)
+         * @type {string|null}
+         */
+        let indexHint = null;
+        /**
+         * @type {int | null}
+         */
+        let cursorBatchSize = null;
+        // run the query and get the results
+        if (query !== null) {
+            // Now run the query to get a cursor we will enumerate next
+            let cursorQuery = await collection.find(query, options).maxTimeMS(maxMongoTimeMS);
+
+            // set batch size if specified
+            if (env.MONGO_BATCH_SIZE || args['_cursorBatchSize']) {
+                cursorBatchSize = args['_cursorBatchSize'] ? parseInt(args['_cursorBatchSize']) : parseInt(env.MONGO_BATCH_SIZE);
```

Comment:
Would 'args['_cursorBatchSize']' ever end up being <= 0? And 'env.MONGO_BATCH_SIZE' be <=0? If not, then the 'if' statement in line 277 might be redundant.